csrc/val_graph_visitor.cpp

```diff
@@ -0,0 +1,118 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <val_graph_visitor.h>
+
+#include <id_model/to_string.h>
+
+namespace nvfuser {
+
+void ValGraphVisitor::traverse() {
+  const ValGroups terminating_inputs = graph().getTerminatingInputs();
+  ValGroups to_visit_ids = terminating_inputs;
+  ValGroups visited_ids;
+
+  ExprGroups to_visit_exprs;
+  ExprGroups visited_exprs;
+
+  auto is_expr_ready = [&](const ExprGroup& expr_group) -> bool {
+    const auto inp_groups = graph().inputGroups(expr_group);
+    return std::all_of(
+        inp_groups.begin(), inp_groups.end(), [&](ValGroup id_group) {
+          return visited_ids.has(id_group) || id_group->empty();
+        });
+  };
+
+  // If any input of the def expr is mapped with the val
+  // group itself, i.e., a trivial expr, allow visiting the
+  // val group first. The trivial expr group will be visited
+  // after the val group.
+  //
+  // Example:
+  //
+  // [i0, 1]
+  // merge
+  // [i0*1]
+  // map i0 and i0*1
+  // ValGroups: {{i0, i0*1}, {1}}
+  //
+  // Then, {i0, i0*1} and {1} would be visited first, then the merge
+  // expr group would be visited. {i0, i0*1} is also an output group
+  // of the merge but since it's already in the visited set, it would
+  // not be visited again.
+  //
+  // See also IdModelTest.ValGraphStmtSort3 for a concrete example.
+  auto is_val_ready = [&](const ValGroup& val_group) -> bool {
+    const ExprGroups& unique_defs = graph().getDefinitions(val_group);
```

Comment:
Can you remind me when a `val_group` can have multiple defining `ExprGroup`s?

Reply:
For example, suppose we have a 1D tensor and a 2D tensor, `t0` and `t1` and:

```
t0: [i0]
t1: [i1, i2, i3]

t2 = reshape(t0, {i4, i5})
// t2: [i4, i5]

t3 = reshape(t1, {i6, i7})
// t3: [i6, i7]

t4 = add(t2, t3)
// t4: [i8, i9]
```

From `t4 = add(t2, t3)`, we would create a val group of `{i4, i6, i8}` as well as `{i5, i7, i9}`. 

The definition of `i4` should be a split of `i0`, whereas `i6` may be something like merge of `i1`, `i2`, `i3` and then split. So the defining expression of `i6` is different from that of `i4`. And since `i4` and `i6` compose a single group, `{i4, i6, i8}`, that group has multiple different definitions. Same is true for the group of `{i5, i7, i9}`.